3) Some preliminary analysis / plots / stats that you can provide at this point to illustrate interesting features of that data. These don't need to be complicated at this point. You can also tell us what type of data you're working with (numeric, nominal, text, images), how much missing data your dataset seems to have, etc.

In [3]:
import pandas as pd
df = pd.read_csv('./us_perm_visas.csv', low_memory=False)

In [4]:
df.head()

add_these_pw_job_title_9089 agent_city agent_firm_name agent_state  \
0                         NaN        NaN             NaN         NaN   
1                         NaN        NaN             NaN         NaN   
2                         NaN        NaN             NaN         NaN   
3                         NaN        NaN             NaN         NaN   
4                         NaN        NaN             NaN         NaN   

  application_type        case_no case_number case_received_date case_status  \
0             PERM  A-07323-97014         NaN                NaN   Certified   
1             PERM  A-07332-99439         NaN                NaN      Denied   
2             PERM  A-07333-99643         NaN                NaN   Certified   
3             PERM  A-07339-01930         NaN                NaN   Certified   
4             PERM  A-07345-03565         NaN                NaN   Certified   

  class_of_admission              ...               ri_pvt_employment_firm_to  \
0                J-1              ...                                     NaN   
1                B-2              ...                                     NaN   
2               H-1B              ...                                     NaN   
3                B-2              ...                                     NaN   
4                L-1              ...                                     NaN   

  ri_us_workers_considered schd_a_sheepherder     us_economic_sector  \
0                      NaN                NaN                     IT   
1                      NaN                NaN  Other Economic Sector   
2                      NaN                NaN              Aerospace   
3                      NaN                NaN  Other Economic Sector   
4                      NaN                NaN           Advanced Mfg   

  wage_offer_from_9089 wage_offer_to_9089 wage_offer_unit_of_pay_9089  \
0              75629.0                NaN                          yr   
1              37024.0                NaN                          yr   
2              47923.0                NaN                          yr   
3                10.97                NaN                          hr   
4             100000.0                NaN                          yr   

  wage_offered_from_9089 wage_offered_to_9089  wage_offered_unit_of_pay_9089  
0                    NaN                  NaN                            NaN  
1                    NaN                  NaN                            NaN  
2                    NaN                  NaN                            NaN  
3                    NaN                  NaN                            NaN  
4                    NaN                  NaN                            NaN  

[5 rows x 154 columns]

In [5]:
df.tail()

add_these_pw_job_title_9089   agent_city  \
374357                         NaN   Buena Park   
374358                         NaN      Seattle   
374359                         NaN   Schaumburg   
374360                         NaN  LOS ANGELES   
374361                         NaN      Phoenix   

                                          agent_firm_name agent_state  \
374357                           Law Offices of Yohan Lee          CA   
374358                         MacDonald Hoague & Bayless          WA   
374359  International Legal and Business Services Grou...          IL   
374360                        LAW OFFICES OF JAMES S HONG          CA   
374361            Fragomen, Del Rey, Bernsen & Loewy, LLP          AZ   

       application_type case_no    case_number case_received_date case_status  \
374357              NaN     NaN  A-16363-85407         2016-12-29   Withdrawn   
374358              NaN     NaN  A-16271-56745         2016-12-30   Withdrawn   
374359              NaN     NaN  A-16354-82345         2016-12-30   Withdrawn   
374360              NaN     NaN  A-16357-84250         2016-12-30   Withdrawn   
374361              NaN     NaN  A-16279-59292         2016-12-30   Withdrawn   

       class_of_admission              ...                \
374357                NaN              ...                 
374358                L-1              ...                 
374359               H-1B              ...                 
374360                B-2              ...                 
374361               H-1B              ...                 

       ri_pvt_employment_firm_to ri_us_workers_considered schd_a_sheepherder  \
374357                       NaN                      NaN                  N   
374358                       NaN                        Y                  N   
374359                       NaN                      NaN                  N   
374360                       NaN                      NaN                  N   
374361                       NaN                      NaN                  N   

       us_economic_sector wage_offer_from_9089 wage_offer_to_9089  \
374357                NaN            28,330.00                NaN   
374358                NaN           122,000.00         142,000.00   
374359                NaN            79,082.00          79,082.00   
374360                NaN            37,981.00                NaN   
374361                NaN            84,926.00         121,500.00   

       wage_offer_unit_of_pay_9089 wage_offered_from_9089  \
374357                        Year                    NaN   
374358                        Year                    NaN   
374359                        Year                    NaN   
374360                        Year                    NaN   
374361                        Year                    NaN   

       wage_offered_to_9089  wage_offered_unit_of_pay_9089  
374357                  NaN                            NaN  
374358                  NaN                            NaN  
374359                  NaN                            NaN  
374360                  NaN                            NaN  
374361                  NaN                            NaN  

[5 rows x 154 columns]

In [8]:
print (len(df))

374362


There are total 374362 number of records and 154 features in this dataset. This dataset includes numeric and nominal values.

In [10]:
print ('Below is the list of features:')
print(df.columns.values)

Below is the list of features:
['add_these_pw_job_title_9089' 'agent_city' 'agent_firm_name'
 'agent_state' 'application_type' 'case_no' 'case_number'
 'case_received_date' 'case_status' 'class_of_admission'
 'country_of_citizenship' 'country_of_citzenship' 'decision_date'
 'employer_address_1' 'employer_address_2' 'employer_city'
 'employer_country' 'employer_decl_info_title' 'employer_name'
 'employer_num_employees' 'employer_phone' 'employer_phone_ext'
 'employer_postal_code' 'employer_state' 'employer_yr_estab'
 'foreign_worker_info_alt_edu_experience'
 'foreign_worker_info_birth_country' 'foreign_worker_info_city'
 'foreign_worker_info_education' 'foreign_worker_info_education_other'
 'foreign_worker_info_inst' 'foreign_worker_info_major'
 'foreign_worker_info_postal_code' 'foreign_worker_info_rel_occup_exp'
 'foreign_worker_info_req_experience' 'foreign_worker_info_state'
 'foreign_worker_info_training_comp' 'foreign_worker_ownership_interest'
 'foreign_worker_yr_rel_edu_complete

In [28]:
#check for missing data
features_missing_counts = pd.Series(0, index=list(df))
for feature in df:
    features_missing_counts[feature] = df[feature].isna().sum()
print ('Missing values for each feature:')
print (features_missing_counts)

Missing values for each feature:
add_these_pw_job_title_9089               332168
agent_city                                162908
agent_firm_name                           167365
agent_state                               166123
application_type                          239093
case_no                                   239093
case_number                               135269
case_received_date                        135271
case_status                                    0
class_of_admission                         22845
country_of_citizenship                     20633
country_of_citzenship                     353788
decision_date                                  0
employer_address_1                            42
employer_address_2                        149193
employer_city                                 14
employer_country                          135343
employer_decl_info_title                  135308
employer_name                                 12
employer_num_employees              

In [29]:
print (features_missing_counts.idxmax() + ': ' + str(features_missing_counts['orig_file_date']))

orig_file_date: 374081


orig_file_date has most missing values.

In [50]:
import numpy as np
petitionByEmployer=df.groupby('employer_name').size()
type(petitionByEmployer)
sortedPetitionByEmployer = petitionByEmployer.sort_values(ascending=False)

In [51]:
print ('Top 10 employer with petition:')
sortedPetitionByEmployer.head(10)                        

Top 10 employer with petition:


employer_name
COGNIZANT TECHNOLOGY SOLUTIONS US CORPORATION    12081
MICROSOFT CORPORATION                             8369
INTEL CORPORATION                                 6564
GOOGLE INC.                                       5165
AMAZON CORPORATE LLC                              3719
INFOSYS LTD.                                      2717
CISCO SYSTEMS, INC.                               2634
ORACLE AMERICA, INC.                              2584
APPLE INC.                                        2583
QUALCOMM TECHNOLOGIES INC.                        1710
dtype: int64